### PR TITLE
Expose MemoryMarshal.AsRef

### LINKED ISSUE #30613: [API Proposal] API for interpretting Span<byte> as struct without copying
MemoryMarshal contains several APIs that allow reinterpreting `Span<byte>`, but none of them are a good fit for interpreting Span<byte> as struct without copying:

- `MemoryMarshal.Read` allows interpreting `Span<byte>` as arbitrary struct. It creates a local copy that is fine and beneficial for small structs, but becomes expensive for large structs.
- `MemoryMarshal.Cast` allows interpreting `Span<byte>` as sequence of arbitrary structs. It does not create a local copy, but comes with extra overhead required to compute the length of the resulting span.
- `MemoryMarshal.GetReference` + `Unsafe.As` does not create a local copy, but also skips any bounds checks that is not good for security defense in depth.

## Proposal

Add methods that allow interpreting `Span<byte>` as arbitrary large struct without copying to MemoryMarshal. 

```csharp
static class MemoryMarshal
{
    public static ref readonly T AsRef<T>(System.ReadOnlySpan<byte> span) where T : struct { throw null; }
    public static ref T AsRef(System.Span<byte> span) where T : struct { throw null; }
}
```

Implementation of these methods is going to be faster functional equivalent of `ref MemoryMarshal.Cast<byte, T>(span)[0]`.

These APIs come with the same set of portability caveats as `Memory.Cast`. They can be only used when the memory block is sufficiently aligned by other means; or  when the target platform does not require strong memory alignment.

### PR DESCRIPTION
**Pull request by jkotas**
- Add tests
- Take advantage of the new API in CoreFX

Fixes #30613

### CODE REVIEW

File: src/System.Memory/ref/System.Memory.cs [hunk 1]
```diff
@@ -538,6 +538,8 @@ public static partial class MemoryMarshal
         public static bool TryRead<T>(System.ReadOnlySpan<byte> source, out T value) where T : struct { throw null; }
         public static bool TryWrite<T>(System.Span<byte> destination, ref T value) where T : struct { throw null; }
         public static void Write<T>(System.Span<byte> destination, ref T value) where T : struct { }
+        public static ref T AsRef<T>(Span<byte> span) where T : struct { throw null; }
```
Comment (ahsonkhan) [inline src/System.Memory/ref/System.Memory.cs]: nit: sort order, place after AsMemory (or auto-generate the ref).

Also, System.{ReadOnly}Span instead.

File: src/System.Memory/tests/System.Memory.Tests.csproj [hunk 2]
```diff
@@ -9,6 +9,8 @@
     <!-- Tests specific to the fast span -->
     <Compile Include="MemoryMarshal\CreateSpan.cs" />
     <Compile Include="MemoryMarshal\CreateReadOnlySpan.cs" />
+    <Compile Include="MemoryMarshal\AsRef.cs" />
```
Comment (ahsonkhan) [inline src/System.Memory/tests/System.Memory.Tests.csproj]: same nit here: alphabetical order

File: src/System.Memory/tests/MemoryMarshal/AsReadOnlyRef.cs [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+
+namespace System.SpanTests
+{
+    public static partial class MemoryMarshalTests
+    {
+        [Fact]
+        public static void AsReadOnlyRef()
+        {
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(new byte[] { 0x11, 0x22, 0x22, 0x11 });
+            ref readonly int asInt = ref MemoryMarshal.AsRef<int>(span);
```
Comment (ahsonkhan) [inline src/System.Memory/tests/MemoryMarshal/AsReadOnlyRef.cs]: How about adding a test case for interpreting a custom struct (both for success and failure, where the struct contains a reference type)?

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj [hunk 4]
```diff
@@ -7,7 +7,6 @@
     <WarningsNotAsErrors>CS0618</WarningsNotAsErrors>
     <DefineConstants>$(DefineConstants);NETCORE</DefineConstants>
     <DefineConstants Condition="'$(TargetsWindows)' == 'true'">$(DefineConstants);FEATURE_WINDOWS_SYSTEM_COLORS;FEATURE_SYSTEM_EVENTS</DefineConstants>
-    <DefineConstants Condition="'$(TargetsUnix)' == 'true'">$(DefineConstants);CORECLR</DefineConstants>
```
Comment (ahsonkhan) [inline src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Just to confirm, these constants were unused prior to this change of adding and using AsRef, correct?

File: src/System.Memory/tests/MemoryMarshal/AsRef.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+using System.Runtime.InteropServices;
+
+namespace System.SpanTests
+{
+    public static partial class MemoryMarshalTests
+    {
+        [Fact]
+        public static void AsRef()
+        {
+            Span<byte> span = new Span<byte>(new byte[] { 0x11, 0x22, 0x22, 0x11 });
```
Comment (ahsonkhan) [inline src/System.Memory/tests/MemoryMarshal/AsRef.cs]: nit: `Span<byte> span = new byte[] { 0x11, 0x22, 0x22, 0x11 };`

File: src/System.Memory/tests/MemoryMarshal/AsReadOnlyRef.cs [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+using System.Runtime.InteropServices;
+
+namespace System.SpanTests
+{
+    public static partial class MemoryMarshalTests
+    {
+        [Fact]
+        public static void AsReadOnlyRef()
+        {
+            ReadOnlySpan<byte> span = new ReadOnlySpan<byte>(new byte[] { 0x11, 0x22, 0x22, 0x11 });
```
Comment (ahsonkhan) [inline src/System.Memory/tests/MemoryMarshal/AsReadOnlyRef.cs]: same nit here, either use var or:
`ReadOnlySpan<byte> span = new byte[] { 0x11, 0x22, 0x22, 0x11 };`

File: src/System.Drawing.Common/src/System.Drawing.Common.csproj (hunk 4 above)
Comment (jkotas) [reply src/System.Drawing.Common/src/System.Drawing.Common.csproj]: Correct.

**Comment by jkotas** [comment]
@dotnet-bot test Windows x86 Release Build  please

**Pull request merged**